### PR TITLE
drivers: eSPI: add eSPI driver support in NPCX7 series

### PR DESCRIPTION
**Pull request by MulinChao**
This PR introduces the Intel Enhanced Serial Peripheral Interface (eSPI) support by three major CLs.

The first one introduces eSPI bus device in npcx7. It includes eSPI device tree declarations, Zephyr eSPI api implementation, OOB (Out of Band tunneled SMBus) support, and configuration files for eSPI test suites in samples/drivers/espi folder.

The second CL contains the drivers of NPCX Host Sub-Modules that serve as an interface between the Host and Core domains. It includes:
- Keyboard and Mouse Controller (KBC) interface.
- Power Management (PM) channels.
- Shared Memory mechanism (SHM).
- Core Access to Host Modules (C2H).
- Mobile System Wake-Up functions (MSWC).
- Serial Port (Legacy UART)

The tasks in the application layer such as 8042, ACPI, and host command can interact with this driver by connecting api or callback functions.

The last one adds a host command peripheral devices for npcx7 series. By add interact api implementation in the hardware bus driver such as ec_host_cmd_periph_npcx_data_received(), the EC host command framework can send and receive data on the npcx7 EC eSPI bus.

Currently, this PR has passed:
- VW handshaking of eSPI test suite.
- OOB transaction through get_pch_temp() function of eSPI test suite.
- port 80 callback function of eSPI test suite.
- KBC IBF/OBE interrupt notification
- ACPI/Host Command (PM Channels) IBF interrupt notification
- EC Host command framework basic test. (We didn't enable the EC_HOST_CMD subsystem by default. Once host sent GET_BOARD_VER command (0x06) request, the main routine in eSPI test suite sent the fixed response data of it directly.)

All the tests are base on Nuvoton host emulator which is an FPGA board and can generate eSPI transactions by the script.

### CODE REVIEW

**Review by jettr — CHANGES_REQUESTED** [review]
I only had a change to look at the first commit. Will look at the other 3 tomorrow morning.

File: drivers/espi/Kconfig.npcx
Comment (MulinChao) [reply drivers/espi/Kconfig.npcx:16]: I follow what Intel did in https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/espi/Kconfig.xec#L14. I found 'config ESPI_OOB_CHANNEL' doesn't declare 'default' property in https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/espi/Kconfig#L66. Is it still an overriding case?

File: drivers/espi/Kconfig.npcx
Comment (MulinChao) [reply drivers/espi/Kconfig.npcx:33]: Will move to the project configuration file in samples/drivers/espi

Comment (MulinChao) [reply drivers/espi/Kconfig.npcx:33]: Done.

File: drivers/espi/espi_npcx.c
Comment (MulinChao) [reply drivers/espi/espi_npcx.c:30]: Done.

File: drivers/espi/espi_npcx.c
Comment (MulinChao) [reply drivers/espi/espi_npcx.c:83]: It is bit order for 32 bits register. I will replace uint32_t with uint8_t.

Comment (MulinChao) [reply drivers/espi/espi_npcx.c:83]: Done.

File: drivers/espi/espi_npcx.c
Comment (MulinChao) [reply drivers/espi/espi_npcx.c:91]: I'd like to keep the register bit name as the same as our datasheet. Or it will hard to find it in the datasheet and divergence issue with our HAL.

File: drivers/espi/espi_npcx.c
Comment (MulinChao) [reply drivers/espi/espi_npcx.c:114]: No, it doesn't

File: drivers/espi/espi_npcx.c
Comment (MulinChao) [reply drivers/espi/espi_npcx.c:216]: Done.

File: dts/arm/nuvoton/npcx/npcx7-espi-vws-map.dtsi
Comment (MulinChao) [reply dts/arm/nuvoton/npcx/npcx7-espi-vws-map.dtsi:12]: I still suggest using Master-to-Slave which also is indicated in our datasheet. The input signal means this signal is sent by bus master (PCH) to bus slave, (EC)

File: soc/arm/nuvoton_npcx/common/reg/reg_def.h
Comment (MulinChao) [reply soc/arm/nuvoton_npcx/common/reg/reg_def.h:382]: For register bit definition, we only declare bit order in reg_def.h and use BIT(x) in driver instance. I'd like to keep it for consistency. If it's crucial feature must meet, I suggest to create a refactor PR after all npcx driver land.

File: soc/arm/nuvoton_npcx/common/soc_dt.h
Comment (MulinChao) [reply soc/arm/nuvoton_npcx/common/soc_dt.h:33]: It extends an array defined from clocks property in a device tree file by UTIL_LISTIFY macro. Please refer https://github.com/zephyrproject-rtos/zephyr/blob/master/include/sys/util.h#L617 for more detail.

File: soc/arm/nuvoton_npcx/common/soc_dt.h
Comment (MulinChao) [reply soc/arm/nuvoton_npcx/common/soc_dt.h:132]: The index is the property of miwu device defined in device tree file. Please refer https://github.com/zephyrproject-rtos/zephyr/blob/master/dts/bindings/interrupt-controller/nuvoton%2Cnpcx-miwu.yaml#L14

File: soc/arm/nuvoton_npcx/common/soc_dt.h
Comment (MulinChao) [reply soc/arm/nuvoton_npcx/common/soc_dt.h:162]: Please see the above.

File: drivers/espi/espi_utils.h
Comment (MulinChao) [reply drivers/espi/espi_utils.h:23]: No, it won't since the larger CL used this function. But I noticed this PR cannot pass the "socketpair" test suite. The root cause is wrong function names for eSPI flash APIs. Hence, I added the fix and raised up the order of this CL. I'm also curious why the other eSPI driver CL won't encounter this problem?

File: soc/arm/nuvoton_npcx/common/soc_dt.h
Comment (MulinChao) [reply soc/arm/nuvoton_npcx/common/soc_dt.h]: Simon also reminded me in the other PR. In the beginning, I think it's unnecessary because they are only used for npcx series internally. But it seems to give reviewers a headache without detailed comment. I will add a CL to add comments to describe parameter, return type, and usage of each macro function.

Comment (MulinChao) [reply soc/arm/nuvoton_npcx/common/soc_dt.h]: Done.

File: drivers/espi/espi_handlers.c
Comment (albertofloyd) [inline drivers/espi/espi_handlers.c:58]: This looks correct as is now, no need to changed since recently fixed by @ceolin

File: samples/drivers/espi/prj_npcx7m6fb_evb.conf
Comment (albertofloyd) [inline samples/drivers/espi/prj_npcx7m6fb_evb.conf]: Unmerged problem

File: drivers/espi/Kconfig
Comment (albertofloyd) [inline drivers/espi/Kconfig:115]: Is this different than CONFIG_ESPI_PERIPHERAL_HOST_IO?

File: soc/arm/nuvoton_npcx/common/reg/reg_def.h
Comment (jettr) [reply soc/arm/nuvoton_npcx/common/reg/reg_def.h:382]: We can keep it for consistent. Just a small comment

File: drivers/espi/espi_npcx.c
Comment (jettr) [reply drivers/espi/espi_npcx.c:91]: This is just putting an underscore between a number and text. Since this is a macro anyway, you won't be able to look up this up in the datasheet since 3IE doesn't exist in code anywhere.

Either way, it is not that big of a deal, so you can leave it the way it is; it just seems like a more consistent name with that extra underscorce

File: drivers/espi/espi_utils.h
Comment (jettr) [reply drivers/espi/espi_utils.h:23]: I think this should be the parent of your larger CL. It can stand on its own and fixes a potential problem. That way your larger eSPI commit won't have any test failures.

File: drivers/ec_host_cmd_periph/ec_host_cmd_npcx.c
Comment (jettr) [inline drivers/ec_host_cmd_periph/ec_host_cmd_npcx.c:5]: This should be an eSPI driver that works for all eSPI implementations (versus a npcx implementation). NPCX should have an eSPI implementation, but the host cmd periph should be based on a bus like eSPI, SPI, or I2C.

I also believe that I need to update the ec_host_cmd_periph API to include a blocking `read` API function. I used semaphores in the initial API instead, but now that I am implementing the `SPI` version of this periph driver, I realize it make things more complicated than it needs to be.

I have a *very* rough draft of the `SPI` implementation for this driver that I [uploaded](https://github.com/jettr/zephyr/tree/spi_periph_drat) to show you what I mean.

File: samples/drivers/espi/prj_npcx7m6fb_evb.conf
Comment (MulinChao) [reply samples/drivers/espi/prj_npcx7m6fb_evb.conf]: Oops. I fixed it in the other CL. Will change the order of this.

File: drivers/espi/Kconfig
Comment (MulinChao) [reply drivers/espi/Kconfig:115]: No, this config is to enable the support of 'EC host command framework'. Please https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/ec_host_cmd_periph/Kconfig#L6 for more detail.

File: drivers/espi/espi_handlers.c
Comment (MulinChao) [reply drivers/espi/espi_handlers.c:58]: Ok. I will get rid of this line.

Comment (MulinChao) [reply drivers/espi/espi_handlers.c]: Done.

File: drivers/espi/Kconfig
Comment (MulinChao) [reply drivers/espi/Kconfig:115]: Done

File: drivers/espi/espi_npcx.c
Comment (MulinChao) [reply drivers/espi/espi_npcx.c:91]: Thanks. What I mean it's the result of this macro is mapping to register bit name in "reg_def.h". (ie. NPCX_ESPIIE_CFGUPDIE -is composed by Chip family + register name + register bit/field name)  If we want to apply this, it also need to modify the register bit name in reg_def.h.

File: drivers/espi/espi_utils.h
Comment (MulinChao) [reply drivers/espi/espi_utils.h:23]: Done.

File: soc/arm/nuvoton_npcx/common/soc_dt.h
Comment (MulinChao) [reply soc/arm/nuvoton_npcx/common/soc_dt.h]: I submitted a CL to add more detail and explanations for all macro functions in this PR. Please take a look.

File: drivers/espi/espi_utils.h
Comment (MulinChao) [reply drivers/espi/espi_utils.h:23]: Done.

File: drivers/ec_host_cmd_periph/ec_host_cmd_npcx.c
Comment (MulinChao) [reply drivers/ec_host_cmd_periph/ec_host_cmd_npcx.c:5]: That's my original plan, too. eSPI driver offers  espi_api_lpc_read_request, espi_api_lpc_write_request  and callbacks to handle the events from host sub-modules such as KBC, APCI IO ports and so on. The Chromium host cmd framework needs a event from the IO port and handle massive data via shared memory or mailbox. 

But the reality is eSPI api functions only support 8041 so far. Please refer  https://github.com/zephyrproject-rtos/zephyr/blob/master/include/drivers/espi.h#L212 . We need to add more op codes after the interface between host commands and espi peripheral is finalized. In order to test the hardware behavior on npcx7 first, this CL follows what you did fo the simulator and passed the sanity test on my side 

Should I drop this commit in this PR? After all host_cmd_periph_xxx api functions are landed, we can create the other PR to support host command framework.

File: drivers/espi/espi_handlers.c
Comment (MulinChao) [reply drivers/espi/espi_handlers.c]: No, it still doesn't work and block this PR. Please refer https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/espi/espi_handlers.c#L58. Might I ask which PR fix it? If not, I will add it back in this CL.

File: drivers/espi/Kconfig.npcx
Comment (jettr) [reply drivers/espi/Kconfig.npcx:16]: Why providing the test after bool, it makes it overrideable

Oh, you are trying to select the existing Kconfig option. I see. I though you were defining a new options. Can you add a comment saying that we are defaulting the existing option to y since that is the typical use case

File: drivers/ec_host_cmd_periph/ec_host_cmd_npcx.c
Comment (jettr) [reply drivers/ec_host_cmd_periph/ec_host_cmd_npcx.c:5]: Let's drop the host command peripheral for now then and we can add it when eSPI and the host command framework are both in a better state

Comment (MulinChao) [reply drivers/ec_host_cmd_periph/ec_host_cmd_npcx.c]: Got it. I will drop it in this PR.

Comment (MulinChao) [reply drivers/ec_host_cmd_periph/ec_host_cmd_npcx.c]: Done.

File: drivers/espi/Kconfig.npcx
Comment (MulinChao) [reply drivers/espi/Kconfig.npcx:17]: Done.

File: samples/drivers/espi/prj_npcx7m6fb_evb.conf
Comment (jettr) [inline samples/drivers/espi/prj_npcx7m6fb_evb.conf:7]: We shouldn't have this config option yet (since it isn't implement yet, right)?
```suggestion
```

**Review by jettr — COMMENTED** [review]
this looks good to me (with minor comment). I have request that other also review this PR. I also know that we will need to wait until the 2.5 merge window open to land this.

File: samples/drivers/espi/prj_npcx7m6fb_evb.conf
Comment (MulinChao) [reply samples/drivers/espi/prj_npcx7m6fb_evb.conf]: Oops. I miss this line when I dropped the implementation CL. Thanks.

Comment (MulinChao) [reply samples/drivers/espi/prj_npcx7m6fb_evb.conf]: Done.

**Review by jettr — APPROVED** [review]

File: soc/arm/nuvoton_npcx/npcx7/Kconfig.defconfig.series
Comment (sjg20) [inline soc/arm/nuvoton_npcx/npcx7/Kconfig.defconfig.series:42]: This doesn't help much. Ideas to expand this:

- some mention of what eSPI is for
- what features the driver supports
- link to eSPI spec

File: soc/arm/nuvoton_npcx/npcx7/soc.c
Comment (sjg20) [inline soc/arm/nuvoton_npcx/npcx7/soc.c:56]: Can we use lower-case hex everywhere?

**Review by sjg20 — APPROVED** [review]

File: soc/arm/nuvoton_npcx/npcx7/Kconfig.defconfig.series
Comment (MulinChao) [reply soc/arm/nuvoton_npcx/npcx7/Kconfig.defconfig.series:42]: Okay, I know eSPI is not a peripheral interface well known. I will add its spec link of Interl website.

Comment (MulinChao) [reply soc/arm/nuvoton_npcx/npcx7/Kconfig.defconfig.series]: Done.

File: soc/arm/nuvoton_npcx/npcx7/soc.c
Comment (MulinChao) [reply soc/arm/nuvoton_npcx/npcx7/soc.c]: Done.

File: dts/bindings/espi/nuvoton,npcx-espi.yaml
Comment (galak) [inline dts/bindings/espi/nuvoton,npcx-espi.yaml:33]: got `required: true` twice here.

Comment (MulinChao) [reply dts/bindings/espi/nuvoton,npcx-espi.yaml]: Thanks for reminding. I will update it.

Comment (MulinChao) [reply dts/bindings/espi/nuvoton,npcx-espi.yaml]: Done.

**Pull request merged**
